## Appendix 1: Single cell RNA sequencing

Single cell RNASeq (scRNASeq) allows the generation of separate sequencing libraries for individual cells in a sample. 10x Genomics is one of the major scRNASeq companies today, and they use Chromium technology which uses a special machine to generate barcoded cDNA samples. This cDNA prep procedure is summarised in [this Youtube video by 10x Genomics](https://www.youtube.com/watch?v=XTZD1eNoRtc&list=PLfaSRwcfHcq2boegRa_WZBWl3r-nD0zb4&index=6&t=0s) (_relevant bit starts at 0:51, sound not required_) and this graphic:

![_Overview of single-cell barcoding procedure with 10x Genomics chromium technology_](https://static.projects.iq.harvard.edu/files/styles/os_files_xxlarge/public/bauercore/files/scv2protocol.png?m=1510953949&itok=7u8ANWwg)

Essentially, the Chromium machine passes barcoded beads through a capillary tube, where they intermingle with single cells (in suspension) and a reverse transcription mastermix. The beads bind the cells. As the mixture travels further down the tube, they encounter an oil mix which causes the bead-cell combos to be encapsulated by a little bubble (or a "reaction vesicle" called Gel Beads in Emulsion or GEMs) which contains the enzymatic master mix. These bubbles then undergo reverse transcription to make barcoded cDNA, and then the oil is removed, releasing the cDNA from the GEMs. Then the barcoded cDNA can undergo normal Illumina sequencing. However, the sequencing only targets the 3' end of each transcript.

It is important to note that scRNASeq does NOT produce 100% coverage of all cells in a sample or all genes in a cell. The current 10x Genomics estimate is that the "capture efficiency" (% of sequenced cells) is 65% and that (depending on which reagent chemistry is used - v1, v2 or v3) "approximately 6.7-8.1%  (v1), 14-15% (v2), or 30-32% (v3) of mRNA transcripts are captured per cell". However, because you are sequencing heaps of cells, and because "detection of a transcript is a stochastic event", "the majority of the transcriptome is captured as a result of sequencing multiple single cells from the same population" ([10x genomics FAQ](https://kb.10xgenomics.com/hc/en-us/articles/360001539051-What-fraction-of-mRNA-transcripts-are-captured-per-cell-)). This phenomenon, whereby the scRNASeq data only captures a small fraction of the transcription of each cell, is also known as "dropout".

(Note that the human genome contains 32k protein-coding and 50k non-coding genes [according to Quan]. [According to Ensembl](http://asia.ensembl.org/Mus_musculus/Info/Annotation), the mouse genome contains 22508 coding genes and 16078 non-coding genes).
